### PR TITLE
Scatter: support bubble and line charts with out-of-order data (alpha)

### PR DESCRIPTION
**Pull request by ryantxu**
This tries a similar approach to #37423

Here is a chart showing 4 dimensions: lat vs lng, color = price, size = count
image

TODO:
- [x] Quadratic size  markers
- [x] Support line options
- [x] ~~Hover color based on display color~~ uniform alpha overlay
- [x] Zoom behavior
- [x] get CI green (uplot changes)

After this PR:
- improve legend -- when based on continuous color
- improve tooltip -- padding issues
- improve config UI

Note, current config UI has weird issue if you toggle between "XY" and "explicit" mode...  the point size config gets added a bunch of times :man_shrugging:

### CODE REVIEW

File: packages/grafana-schema/src/schema/general.cue [hunk 1]
```diff
@@ -0,0 +1,3 @@
+package schema
+
+AutoAlwaysNever:    "auto" | "never" | "always"                             @cuetsy(kind="enum")
```
Comment (ryantxu) [inline packages/grafana-schema/src/schema/general.cue:3]: I will make a new PR with only this change -- it touches lots of files

File: packages/grafana-ui/src/components/uPlot/config/UPlotConfigBuilder.ts [hunk 2]
```diff
@@ -269,5 +278,19 @@ type UPlotConfigPrepOpts<T extends Record<string, any> = {}> = {
   allFrames: DataFrame[];
 } & T;
 
+/** @alpha */
+//type UPlotConfigPrepOptsXY = Omit<UPlotConfigPrepOpts, 'frame' | 'getTimeRange' | 'timeZone'>;
+
+/** @alpha */
+type UPlotConfigPrepOptsXY<T extends Record<string, any> = {}> = {
```
Comment (ryantxu) [inline packages/grafana-ui/src/components/uPlot/config/UPlotConfigBuilder.ts:285]: @leeoniya  still necessary?

**Comment by ryantxu** [comment]
@dprokop -- thoughts on this for 8.2?  the xy plot is all alpha, so that is low risk and no real need for it in 8.2.  The pro *and* con for merging into 8.2 is that this updates the `alignedData` argument in config builder -- if this is merged into 8.2, any backport issues will be much smoother, but obviously not great to add a change with no real impact so :shrug:

File: public/app/plugins/panel/xychart/XYChartPanel2.tsx [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            </UPlotChart>
+          )}
+        </VizLayout>
+        <Portal>
+          {hover && (
+            <VizTooltipContainer position={{ x: hover.pageX, y: hover.pageY }} offset={{ x: 10, y: 10 }}>
+              <TooltipView series={series[hover.scatterIndex]} rowIndex={hover.xIndex} data={data.series} />
+            </VizTooltipContainer>
+          )}
+        </Portal>
```
Comment (dprokop) [inline public/app/plugins/panel/xychart/XYChartPanel2.tsx:122]: This is a complete convention shift, yet I think it's quite elegant still 👍  

My concern is the panel state may become unnecessarily complex and verbose with the entire addons (tooltip, legend etc) lifecycle handling. Is this experimentation only or have you encountered some major problems with TooltipPlugin API alignment? maybe @ryantxu ?

File: public/app/plugins/panel/xychart/ExplicitEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+import { XYChartOptions, ScatterFieldConfig } from './models.gen';
+
+export const ExplicitEditor: FC<StandardEditorProps<ScatterFieldConfig[], any, XYChartOptions>> = ({
+  value,
+  onChange,
+  context,
+}) => {
+  return <div>TODO: explicit scatter config</div>;
+};
```
Comment (dprokop) [inline public/app/plugins/panel/xychart/ExplicitEditor.tsx:12]: Guess we can get rid of this as it's handled by options API.

File: public/app/plugins/panel/xychart/XYChartPanel2.tsx [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    series: [],
+  };
+
+  componentDidMount() {
+    this.initSeries(); // also data
+  }
+
+  componentDidUpdate(oldProps: Props) {
+    const { options, data } = this.props;
+    const configsChanged = options !== oldProps.options || data.structureRev !== oldProps.data.structureRev;
```
Comment (dprokop) [inline public/app/plugins/panel/xychart/XYChartPanel2.tsx:40]: `options !== oldProps.options` - I'm not 100% sure about this one - ain't these two going to be the same in situation when deep options change, i.e. thresholds or value mappings config? My gut feeling is it's gonna be indicated by the structureRev change, but still quite complex. Think this wasn't that visible with the hooks invalidation, which makes me like class component approach better in this case.

File: public/app/plugins/panel/xychart/XYChartPanel2.tsx [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  };
+
+  renderLegend = () => {
+    const { data } = this.props;
+    const { series } = this.state;
+    const items: VizLegendItem[] = [];
+    for (const s of series) {
+      const frame = s.frame(data.series);
+      if (frame) {
+        for (const item of s.legend(frame)) {
```
Comment (dprokop) [inline public/app/plugins/panel/xychart/XYChartPanel2.tsx:80]: 👍

**Review by dprokop — APPROVED** [review]
@ryantxu I went through the code - plenty of stuff to digest, but the overall direction seems good. Left some concerns in comments.

All in all new scatter implementation with the explicit tooltip and legend handling feels like a fundamental change, yet I find it somehow easier to reason about :) Was thinking what will happen if we make this significant decision and apply this pattern in time series, bar chart etc panels. And the conclusion would be it's not going to be that bad as in the end in contexts other than Dashboard we already use PanelRenderer which removes the need for re-implementing stuff like Tooltip etc. 

For beta2 I think this is OK to be merged since it's an alpha panel. I've quickly tested TimeSeries/BarChart if there weren't any regressions, but all works fine. 

XY config still feels rough and unpredictable (couldn't get the color dimension to work for instance if specified via field), changing the XY vs Explicite mode leaves Point size dimension config UI in the panel editor, but these are details we can hammer out later.

Great progress!

**Pull request merged**